### PR TITLE
Use cli services

### PR DESCRIPTION
**Pull request by crookse**
**Description**

* Uses drashland/services CLI services
* DRYs code
* Changes the `rhum` command to the following: `rhum [command] [options] [directory|file]`. previously it was `rhum [options] [directory|file]`
* Commands:
    ```
    rhum make
    rhum test
    rhum help, rhum --help
    rhum version, rhum --version
    ```
* Uses `LoggerService` instead of `logInfo()`, `logError()`, etc.
* Uses one bumper script to handle bumping for pre-release and when deno updates: `console/bumper_ci_service.ts`
* Adds tests (`tests/unit/console/bumper_ci_service_test.ts`) that test what the `bumper.yml` script would do when it runs `console/bumper_ci_service.ts`. In these tests, the file being passed to `BumperService.bump()` is the `tests/data/pattern_types.txt` file. This test file has a number of patterns that the regex in the `BumperService.bump()` method should match and replace.
* Removes duplicated regex patterns. See `console/bumper_ci_service_files.ts`

### CODE REVIEW

File: src/cli/services/make_service.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const encoder = new TextEncoder();
+
+export const testTemplate: Uint8Array = encoder.encode(
+  `import { Rhum } from "https://deno.land/x/rhum@v1.1.4/mod.ts";
```
Comment (ebebbington) [inline src/cli/services/make_service.ts:4]: needs to be references in the update versions script

**Review by ebebbington — APPROVED** [review]
Other than the one comment, lgtm!

**Comment by ebebbington** [comment]
Oh actually another thing i forgot, maybe it's best to have `--version` and `--help` to be consistent? though then again.. it feels like 50/50 most of the time for cli tools

File: src/cli/services/make_service.ts (hunk 1 above)
Comment (crookse) [reply src/cli/services/make_service.ts]: d5eca1bb

Comment (crookse) [reply src/cli/services/make_service.ts]: 6e966a716bda53

Comment (ebebbington) [reply src/cli/services/make_service.ts]: I think we have update Deno versions script for the bumper, so when our deps are updated we update the references too eg Deno version in workflows, then we have bump versions script which is for the pre release, to update the Rhum versions ready for a new release

So might need to move this into the bump versions script - I did add that bump versions thing to services but I believe we could abstract the getting of the versions from the args into that services file, not sure how tho

File: src/cli/commands/test.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      "You cannot use --filter-test-case and --filter-test-suite together. Please specify one option.",
+    );
+    Deno.exit();
+  }
+
+  const testFileOrDir = args[args.length - 1];
+
+  if (testFileOrDir.includes("--")) {
+    logError("Please specify a test file or directory.");
+    Deno.exit();
```
Comment (ebebbington) [inline src/cli/commands/test.ts:28]: Maye `deno.exit(1)` is best here too

File: console/bumper_ci_service.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import { BumperService } from "https://raw.githubusercontent.com/drashland/services/master/ci/bumper_service.ts";
+
+const b = new BumperService("rhum", Deno.args);
+
+if (BumperService.is_for_pre_release) {
+  b.bump([
+    {
+      filename: "./README.md",
+      replaceTheRegex: /dmm@v[0-9\.]+[0-9\.]+[0-9\.]/g,
+      replaceWith: `dmm@v{{ thisModulesLatestVersion }}`,
```
Comment (ebebbington) [inline console/bumper_ci_service.ts:10]: copied from dmm?

File: console/bumper_ci_service.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ if (BumperService.is_for_pre_release) {
     {
       filename: "./src/cli/commands/make.ts",
       replaceTheRegex:
-        `import { Rhum } from "https://deno.land/x/rhum@v1.1.4/mod.ts`,
+        /import \{ Rhum \} from "https\:\/\/deno.land\/x\/rhum\@v[0-0]\.[0-9]\.[0-9]\/mod.ts"/g,
       replaceWith:
         `import { Rhum } from "https://deno.land/x/rhum@v{{ thisModulesLatestVersion }}/mod.ts"`,
```
Comment (ebebbington) [inline console/bumper_ci_service.ts]: tbbh you could just do `replaceTheRegex: /rhum@v[...]/g, replaceWith: `rhum@v{{ ... }}`? or is there a reason to do the whole url

**Review by ebebbington — CHANGES_REQUESTED** [review]

File: src/cli/commands/test.ts (hunk 2 above)
Comment (crookse) [reply src/cli/commands/test.ts]: d9953dd

File: src/cli/commands/test.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    if (!filters.test_suite) {
+      filters.test_suite = getFilterTestSuiteValue(arg);
+    }
+  });
+
+  if (filters.test_case && filters.test_suite) {
+    logError(
+      "You cannot use --filter-test-case and --filter-test-suite together. Please specify one option.",
+    );
+    Deno.exit();
```
[the remark this answers is not in the document]
Comment (crookse) [reply src/cli/commands/test.ts]: d9953dd

File: src/cli/commands/make.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+);
+
+export function make(args: string[]) {
+  let testFileToCreate = args[0];
+
+  if (!testFileToCreate.includes(".ts")) {
+    logError(
+      `Test files require a .ts extension. You passed in "${testFileToCreate}" as the test file.`,
+    );
+    Deno.exit();
```
[the remark this answers is not in the document]
Comment (crookse) [reply src/cli/commands/make.ts]: d9953dd

File: console/bumper_ci_service.ts (hunk 3 above)
Comment (crookse) [reply console/bumper_ci_service.ts]: daa92f87e0

File: src/cli/services/make_service.ts (hunk 1 above)
Comment (crookse) [reply src/cli/services/make_service.ts]: i did find an issue (via tests) with the regex i passed in. it's fixed now though. here's the file:

https://github.com/drashland/rhum/pull/80/files#diff-2e202eda61bc9862e1f12e66fe6fcbbf971e0da38fb65bbaf846eade32aa6116R22-R28

File: console/bumper_ci_service.ts (hunk 4 above)
Comment (crookse) [reply console/bumper_ci_service.ts]: i just did a quick and dirty implementation

Comment (crookse) [reply console/bumper_ci_service.ts]: a26f53046daae47c6b

Comment (crookse) [reply console/bumper_ci_service.ts]: i added tests:

https://github.com/drashland/rhum/pull/80/files#diff-83783ac33429890becf78098debd95619663ffb8e6f1b1906d733d5e0f6710c5

File: tests/unit/console/bumper_ci_service_test.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
+import { Rhum } from "../../../mod.ts";
```
Comment (ebebbington) [inline tests/unit/console/bumper_ci_service_test.ts:1]: mod.ts?

Comment (crookse) [reply tests/unit/console/bumper_ci_service_test.ts:1]: yea i don't see why rhum can reference itself. should it not?

Comment (ebebbington) [reply tests/unit/console/bumper_ci_service_test.ts:1]: im an idiot, i think i thought i was looking at the services repo, so i was confused why a dependency was being imported from mod.ts - you were right to question my stupidity!

**Review by ebebbington — APPROVED** [review]
LGTM

**Pull request merged**